pyodide_build/buildall.py

```diff
@@ -5,79 +5,183 @@
+# circular dependencies, then all packages should have been built, which we
+# check with an assert just to be sure.
+def build_packages(packagesdir, outputdir, args):
+    pkg_map: Dict[str, Package] = {}
+
+    packages: Optional[Set[str]] = common._parse_package_subset(args.only)
+    if packages is None:
+        packages = set(
+            str(x) for x in packagesdir.iterdir() if (x / "meta.yaml").is_file()
+        )
+
+    # Generate Package objects for all specified packages and recursive
+    # dependencies.
+    while packages:
+        pkgname = packages.pop()
+        if pkg_map.get(pkgname) is not None:
             continue
 
-        pkgpath = pkgdir / "meta.yaml"
-        if pkgdir.is_dir() and pkgpath.is_file():
-            pkg = common.parse_package(pkgpath)
-            name = pkg["package"]["name"]
-            reqs = pkg.get("requirements", {}).get("run", [])
-            dependencies[name] = reqs
-            imports = pkg.get("test", {}).get("imports", [name])
-            for imp in imports:
-                import_name_to_package_name[imp] = name
-
-    for pkgname in dependencies.keys():
-        build_package(pkgname, dependencies, packagesdir, outputdir, args)
-
-    # The "test" package is built in a different way, so we hardcode its
-    # existence here.
-    dependencies["test"] = []
+        pkg = Package(packagesdir / pkgname)
+        pkg_map[pkg.name] = pkg
+
+        for dep in pkg.dependencies:
+            # This won't catch all duplicates but let's try our best
+            if pkg_map.get(dep) is None:
+                packages.add(dep)
+
+    # Build set of dependents
+    for pkg in pkg_map.values():
+        for dep in pkg.dependencies:
+            pkg_map[dep].dependents.add(pkg.name)
+
+    # Insert packages into build_queue. We *must* do this after counting
+    # dependents, because the ordering ought not to change after insertion.
+    build_queue: PriorirtyQueue = PriorityQueue()
+    for pkg in pkg_map.values():
+        if len(pkg.dependencies) == 0:
+            build_queue.put(pkg)
+
+    built_queue = Queue()
+
+    def builder(n):
+        print(f"Starting thread {n}")
+        while True:
+            pkg = build_queue.get()
+            print(f"Thread {n} building {pkg.name}")
+            pkg.build(outputdir, args)
+            print(f"Thread {n} built {pkg.name}")
+            built_queue.put(pkg)
+            # Release the GIL so new packages get queued
+            sleep(0.01)
 
+    for n in range(0, args.num_threads):
+        Thread(target=builder, args=(n + 1,), daemon=True).start()
+
+    num_built = 0
+    while num_built < len(pkg_map):
+        pkg = built_queue.get()
+        num_built += 1
+
+        for dependent in pkg.dependents:
+            dependent = pkg_map[dependent]
+            dependent.unbuilt_dependencies.remove(pkg.name)
+            if len(dependent.unbuilt_dependencies) == 0:
+                build_queue.put(dependent)
```

Comment:
I don't think this is something that could happen. A build is triggered only if the number of unbuilt dependencies goes from 1 to 0, which happens exactly once (`set.remove` raises an exception when removing non-existent elements). Since there is only one thread doing this removal and queuing process, there is no risk of race condition.